dynamiqs/qarray.py

```diff
@@ -0,0 +1,554 @@
+
+    def __rsub__(
+        self, other: ScalarLike | ArrayLike
+    ) -> QArray:  # warning if used with scalar
+        """Element-wise subtraction with a scalar or an array on the right."""
+        return -self + other
+
+    @abstractmethod
+    def __matmul__(self, other: ArrayLike) -> QArray:
+        """Matrix multiplication with another quantum state or JAX array."""
+
+    def __rmatmul__(self, other: ArrayLike) -> QArray:
+        """Matrix multiplication with another quantum state or JAX array
+        on the right.
+        """
+
+
+class DenseQArray(QArray):
+    r"""DenseQArray is QArray that uses JAX arrays as data storage."""
+
+    data: ArrayLike
+    dims: tuple[int, ...]
+
+    def __init__(self, data: ArrayLike, dims: tuple[int, ...] | None = None):
+        if not (isbra(data) or isket(data) or isdm(data) or isop(data)):
+            raise ValueError(
+                f'DenseQArray data must be a bra, a ket, a density matrix '
+                f'or and operator. Got array with size {data.shape}'
+            )
+        if dims is None:
+            dims = data.shape[-2] if isket(data) else data.shape[-1]
+            dims = (dims,)
+
+        self.data = data
+        self.dims = dims
+
+    @property
+    def dtype(self) -> jnp.dtype:
+        return self.data.dtype
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        return self.data.shape
+
+    @property
+    @pack_dims
+    def I(self) -> Array:  # noqa: E743
+        return jnp.eye(jnp.prod(jnp.asarray(self.dims)))
+
+    @property
+    @pack_dims
+    def mT(self) -> Array:
+        return self.data.mT
+
+    @pack_dims
+    def conj(self) -> Array:
+        return self.data.conj()
+
+    def norm(self) -> Array:
+        return norm(self.data)
+
+    @pack_dims
+    def unit(self) -> Array:
+        return unit(self.data)
+
+    @pack_dims
+    def diag(self) -> Array:
+        return jnp.diag(self.data)
+
+    def trace(self) -> Array:
+        return jnp.trace(self.data)
+
+    @pack_dims
+    def reshape(self, *new_shape: int) -> Array:
+        return jnp.reshape(self.data, new_shape)
+
+    @pack_dims
+    def broadcast_to(self, *new_shape: int) -> Array:
+        return jnp.broadcast_to(self.data, new_shape)
+
+    def ptrace(self, keep_dims: tuple[int, ...]) -> DenseQArray:
+        return DenseQArray(
+            ptrace(self.data, keep_dims, self.dims),
+            tuple(self.dims[dim] for dim in keep_dims),
+        )
+
+    @pack_dims
+    def dag(self) -> Array:
+        """Dagger of the quantum state.
+
+        Returns:
+            DenseQArray: The dagger of the quantum state.
+        """
+        return dag(self.data)
+
+    def isket(self) -> bool:
+        return isket(self.data)
+
+    def isbra(self) -> bool:
+        return isbra(self.data)
+
+    def isdm(self) -> bool:
+        return isdm(self.data)
+
+    def isherm(self) -> bool:
+        # TODO: could be made more efficient
```

Comment:
I'm not sure it could? This is in O(n^2) and the naive algorithm is probably the best one here.

Reply:
This version does twice the job, the scaling is the good one but we could improve the constant

Reply:
Ah indeed!